### ISSUE TITLE
[ILM] Configuring force merge and searchable snapshot in hot phase

### ISSUE THREAD

**Issue by jloleysens** [elastic/elasticsearch#65730]
## Summary

Setting searchable snapshots in the hot phase disables forcemerge, shrink and a number of other actions in subsequent phases. But there is some confusion on searchable snapshot interacts with forcemerge in the hot phase when translating this to UI/UX terms.

## Assumptions

The injected force merge action by searchable snapshot will be a noop only if a preceding force merge action merged down to 1 segment, otherwise it will merge down to one segment (per [the docs](https://www.elastic.co/guide/en/elasticsearch/reference/master/ilm-searchable-snapshot.html#ilm-searchable-snapshot-options), although the docs indicate that _any_ forcemerge action that preceeds snapshot makes the injected forcemerge a noop). Is this understanding incorrect or are the docs incorrect? CC @andreidan 

## Proposal

At the moment, it is possible to configure an ILM policy with both force merge and searchable snapshot in the hot phase. The force merge action in the policy enables control of segments and compression quality. However, when searchable snapshot is configured a forcemerge action is injected which can create a situation where data goes through two rounds of forcemerge actions given the above assumption.

It would be best if the ILM API would disallow setting forcemerge when searchable snapshot is set. In this way, Elasticsearch will do the actions it knows is best for preparing for searchable snapshots. If users really want to avoid force merging they can also use the "advanced" setting disabling the injected force merge action.

In Kibana, it would best to remove the ability to configure both searchable snapshot and force merge since it reduces the ability to configure policies that are sub-optimal, I think it would be great if we supported this at the API level too unless there is a clear case why we want users to be able configure a separate forcemerge action.

CC @andreidan @dakrone  @jethr0null 

## Addtional

* PR introducing Searchable Snapshots to ILM UI in Kibana https://github.com/elastic/kibana/pull/83783

**Comment by jloleysens** [comment]
Apologies, wrong @leehinman - intended to ping @dakrone . GitHub autocomplete did me in.

**Comment by jethr0null** [comment]
cc @henningandersen

**Comment by andreidan** [comment]
IMO we should be allowing the actions to both be configured. The step in the searchable snapshot action was added to be absolutely sure we forcemerge when mounting the index as searchable snapshot, and it is a prerequisite for mounting the index and not as taking over the responsibility from the `forcemerge` action (ie. if `move-to-step` skips over the `forcemerge` action straight to `searchable_snapshot` the index will be mounted optimally as the injected forcemerge step will make sure of this)

> However, when searchable snapshot is configured a forcemerge action is injected which can create a situation where data goes through two rounds of forcemerge actions given the above assumption.

This is something we are aware of, however if the index was already force merged, the injected `forcemerge` *step* will fly by without performance/time detrimental implications to the policy execution.

> although the docs indicate that any forcemerge action that preceeds snapshot makes the injected forcemerge a noop

That is correct. If the index has 1 segment already (by virtue of a previous `forcemerge` action, or manual force merging) _any_ subsequent forcemerge attempt (action, step, or manual) will be a no-op.

> It would be best if the ILM API would disallow setting forcemerge when searchable snapshot is set. 

This would open questions regarding which force merge and which searchable snapshot actions? Should we disable force merge in hot if searchable snapshot is configured in cold? I'd argue we shouldn't do this as force merge would be faster in the hot tier, and the forcemerge step in cold would be a no-op. 

There'd still be the question if the `searchable_snapshot` and `forcemerge` action within the _same_ phase should invalidate themselves (as `searchable_snapshot` performs a `forcemerge` by default). My opinion is that they shouldn't for 2 reasons:
1) having to reason about cross-actions dependencies builds up the cognitive load (`searchable_snapshot` force merges by default and is a no-op if index was forcemerged already - even outside ILM) 
2) one might still want to forcemerge in hot (to take advantage of the better hardware resources) and searchable snapshot in cold only. This would create add complexity to the validation as we want to allow hot:forcemerge -> cold: searchable_snaphort but we do not want to allow hot:forcemerge, searchablesnapshot

Adding `team-discuss` to see if there are other opinions about this ^.

> In Kibana, it would best to remove the ability to configure both searchable snapshot and force merge since it reduces the ability to configure policies that are sub-optimal,

IMO Kibana should not allow disabling the `forcemerge` step in the searchable snapshot action (which would still be available via the API), irrespective of an explicitly configured `forcemerge` action in the same phase.

**Comment by jloleysens** [comment]
@andreidan Thanks for the explanation.

TL;DR

We drove this to a conclusion offline: configuring searchable snapshot + forcemerge in hot is the behavior we want to preserve in API and at the UX level.

Your response:

> which force merge and which searchable snapshot actions? Should we disable force merge in hot if searchable snapshot is configured in cold? I'd argue we shouldn't do this as force merge would be faster in the hot tier, and the forcemerge step in cold would be a no-op

Helped me understand this better. Thinking about the sequence of possible actions being that forcemerge could happen before searchable snapshot _anyway_ is a good point. I was perhaps overindexing on doing them in the same phase being the issue.

The heart of the issue, for me, was trying to understand how best to present this to users in a more opinionated way. How do we guide users to make the best decisions for their policies? What decisions can we make for them? At the other end of this is making sure we do not create totally different constraints in Kibana vs Elasticsearch. Namely: hiding forcemerge in hot when searchable snapshot is on in hot in Kibana. Technical issues being: Is the forcemerge config secretly still there? Or did we remove it? If we removed it did we throw away user config? What if we load a policy that has both SS and forcemerge configured in hot in the UI? Do we throw away forcemerge without telling the user? These were the situations I was trying to avoid by driving the question of forcemerge and searchable snapshot in hot to a conclusion. Thanks for helping me understand this better!
